Flatten out DiscontinuedService

Jngai opened the pull request:
#1112 

# Steps toward merge (reviewer: @alexsoble)

+ [x] Code review
+ [x] Test data migration against demo data locally
+ [x] Test data migration against demo data on Heroku demo site
+ [x] Demo site/Internet Explorer: Test record/discontinue UI on Internet Explorer, make sure `discontinued_at` and `discontinued_by_educator_id` are set correctly 
+ [x] Test service upload UI (demo site/Chrome — filesystem not working in IE VM)
+ [x] Test Student Report PDF rendering  (demo site/Chrome — filesystem not working in IE VM)
+ [ ] Capture a backup of database before deploying
+ [ ] Deploy and run `heroku run rake services:move_discontinued_at_to_services`

Jngai commented:
@alexsoble I started off rewiring the controllers, did some work in the model and spec. I will be making daily changes. Feel free to start reviewing.

alexsoble commented:
@Jngai OK sounds good!

alexsoble commented:
@Jngai When you're ready for a QA I can download some production data to my machine and run the migration you wrote against it to test.

Jngai commented:
@alexsoble OK I have two more tests that I need to get passed. Afterward I am going to check my work and see what else there is left then will reach out to you.

Jngai commented on app/assets/javascripts/student_profile/services_list.jsx:
```diff
@@ -63,7 +63,7 @@ import _ from 'lodash';
 
     wasDiscontinued: function(service) {
-      return (service.discontinued_by_educator_id !== null);
+      return (service.discontinued_recorded_at !== null);
```
The field `discontinued_at` has replaced `DiscontinuedServices`. If a service is active then its `discontinued_at` will be not `nill`.

Jngai commented on app/controllers/service_uploads_controller.rb:
```diff
@@ -45,15 +45,9 @@ def create
-          discontinued_at: date_ended,
-          recorded_by_educator: current_educator,
-        )
+        service.update_attributes(:discontinued_at => date_ended)
```
I replace `DiscontinuedService.new` with an `update_attributes(:discontinued_at) here and everywhere else.

Jngai commented on app/controllers/services_controller.rb:
```diff
@@ -10,18 +10,15 @@ def authorize!
-      discontinued_at: Time.now
-    })
+    service = Service.find(service_id)
+    service.update_attributes(discontinued_at: Time.now)
```
same with the above except here we use `Time.now` for `discontinued_at`.

Jngai commented on app/controllers/students_controller.rb:
```diff
@@ -203,7 +198,7 @@ def set_up_student_report_data
 
     # Load services for the student for the filtered dates
-    @services = @student.services.includes(:discontinued_services).where("date_started <= ? AND (discontinued_services.discontinued_at >= ? OR discontinued_services.discontinued_at IS NULL)", @filter_to_date, @filter_from_date).order('date_started, discontinued_services.discontinued_at').references(:discontinued_services)
+    @services = @student.services.where("date_started <= ? AND (discontinued_at >= ? OR discontinued_at IS NULL)", @filter_to_date, @filter_from_date).order('date_started, discontinued_at')
```
removed all sql references to `discontinued_services`.

alexsoble replied:
👍

Jngai commented on app/models/service.rb:
```diff
@@ -5,11 +5,12 @@ class Service < ActiveRecord::Base
 
-  validates_presence_of :recorded_by_educator_id, :student_id, :service_type_id, :recorded_at, :date_started#, :estimated_end_date
+  validates_presence_of :recorded_by_educator_id, :student_id, :service_type_id, :recorded_at, :date_started
+  validate :must_be_discontinued_after_service_start_date
```
This validation was originally from the discontinued_service model which I carry over.

Jngai commented on app/models/service.rb:
```diff
@@ -5,11 +5,12 @@ class Service < ActiveRecord::Base
   def discontinued?
-    (discontinued_services.size > 0) && !has_scheduled_end_date?  # If the end date is in the future
-                                                                  # the service isn't discontinued yet.
+    !discontinued_at.nil? && !has_scheduled_end_date? # If the end date is in the future
```
Here @alexsoble please check my work for this model. I refactored a bit. I am just worried that I didn't get the original intent of the method right.

alexsoble replied:
Hey @Jngai! This is a method that will get a lot less complex now that we're flattening out the service model. 

I would say this method could look something like:

```
discontinued_at.present? && (Time.now > discontinued_at)
```

It's also good to try to avoid "double negative" expressions like `!discontinued_at.nil?` and replace them with "positive" expressions like `discontinued_at.present?`.

Jngai commented on lib/tasks/move_discontinued_at_to_services.rake:
```diff
@@ -0,0 +1,19 @@
+  task move_discontinued_at_to_services: :environment do
+    services = Service.all
+    puts "Going to update each service discontinued at"
+
```
A rake task to move `discontinued_at` from `discontinuedservice` to service.

Jngai replied:
and not removing the `discontinued service` model and drop its table until this runs because you need the `has_many` relationship to move data.

alexsoble replied:
Looks correct to me!

Jngai commented on spec/factories/services.rb:
```diff
@@ -4,6 +4,7 @@
     association :service_type
     date_started Date.new(2014, 9, 9)
     recorded_at Date.new(2014, 11, 2)
+    discontinued_at Date.new(2014, 12, 2)
```
adding this to services factory threw the tests off a little bit which I adjusted below.

Jngai commented on spec/models/discontinued_service_spec.rb:
```diff
@@ -1,38 +0,0 @@
-RSpec.describe DiscontinuedService do
-
-  describe '#must_be_discontinued_after_service_start_date' do
-    let(:educator) { FactoryGirl.create(:educator) }
```
I moved this test here to `service_spec.rb` so its good to be deleted.

Jngai commented on spec/factories/discontinued_services.rb:
```diff
@@ -1,7 +0,0 @@
-FactoryGirl.define do
-  factory :discontinued_service do
-    association :service
-    discontinued_at Date.new(2014, 11, 2)
```
no longer necessary

alexsoble replied:
👍

Jngai commented on app/models/service.rb:
```diff
@@ -20,13 +21,21 @@ def end_date
   end
 
+  def must_be_discontinued_after_service_start_date
+    unless discontinued_at.nil?
```
added this extra `unless` statement to make it precise.

Jngai commented:
Hey @alexsoble Will you check this out? I did some work to the `service uploads controller` and the pdf rendering which needs you to try them with an uploading process and a pdf rendering process. I think I removed everything related discontinuedservice. Will you check to see if I miss anything? Also about `takenote.jsx` I think that is not related to discontinuedservice. I could be wrong. Basically every discontinuedservice related functions should be intact despite my refactoring.

Jngai commented:
@alexsoble I added clarifications to my code too.

Jngai commented on app/views/students/student_report.pdf.erb:
```diff
@@ -89,8 +89,8 @@ ul, li {
-            <% if service.discontinued_services.count > 0%>
-              <%= "#{format_date(service.discontinued_services.last.try(:recorded_at))}" %>
+            <% if service.discontinued_at != nil%>
+              <%= "#{format_date(service.try(:recorded_at))}" %>
```
@alexsoble when you test the student report pdf generating process, please make sure the `recorded_at` continues to show.

Jngai commented on app/controllers/service_uploads_controller.rb:
```diff
@@ -45,15 +45,9 @@ def create
 
```
@alexsoble I want to make sure service batch uploads continues to work after these changes.

alexsoble commented on db/schema.rb:
```diff
@@ -265,6 +266,7 @@
     t.string "provided_by_educator_name"
     t.integer "service_upload_id"
     t.datetime "estimated_end_date"
+    t.datetime "discontinued_at"
```
Hey @Jngai! I was expecting to see one more field here, something like `discontinued_by_educator_id`. That's because we want to make sure we're storing a **complete** history of what happens to a service, including which educator discontinued the service. A service could be created by one educator and then discontinued by a second educator.

Jngai commented:
@alexsoble alexsoble I added `discontinued_by_educator_id ` to the schema, made a migration and added it to all `update_attributes`. I mapped it to `current_educator.id`.  I originally thought `recorded_by_educator_id` was enough and it already existed in `services`.  Care to take another look?

alexsoble commented:
@Jngai Yes!

Jngai commented on app/serializers/service_serializer.rb:
```diff
@@ -11,13 +11,13 @@ def serialize_service
       :recorded_at
     ]).merge({
-      discontinued_by_educator_id: discontinued.try(:recorded_by_educator_id),
+      discontinued_by_educator_id: discontinued.try(:discontinued_by_educator_id),
```
@alexsoble I changed this to `discontinued_by_educator_id` as well.

alexsoble replied:
👍

alexsoble commented:
Hey @Jngai! Gave this some manual QA in my local environment. The process of creating, setting an end date for, and discontinuing services worked smoothly. One thing we're missing is a migration to convert existing Services/DiscontinuedServices over to the new "flattened" data structure by copying over the information from each DiscontinuedService onto each service. It's good to add lots of helpful logging statements to these kind of migrations so that we can QA them and make sure they're working as intended. Check out the data migration in #1178 as an example of how to write a migration like this. Let me know if that makes sense to you or if you have any questions, and thanks for tackling this issue @Jngai!

Jngai commented:
Hey @alexsoble Did you see the rake task in the pull request and the temporary added rake task in the deployment script to move existing discontinued services to service?

I am not a big fan of doing data migrations in the migration file. Here is a blog from thoughtbot about it https://robots.thoughtbot.com/data-migrations-in-rails. 

I hope you also checked the uploading process and or every other way that a discontinued service gets used.

Jngai commented:
@alexsoble Its at `lib/tasks/move_discontinued_at_to_services.rake` FYI I didn't try it but should be self explanatory.

Jngai commented:
If you are good with this @alexsoble then please merge it because the current work in master is conflicting with this pull request.

Jngai commented:
@alexsoble take your time to review but longer this stays open, the more conflicts there will be.

alexsoble commented:
@Jngai On the rake task discussion — Thanks for sharing the thoughtbot article! We've mostly done data migrations so far and that's worked well for a single school district running on a single instance. But as we scale up to more districts, this will get less workable for all the reasons the thoughtbot piece describes. So thanks for moving in the direction of rake tasks over migrations and for sharing the article. 👍

alexsoble commented:
@Jngai Things for me to test/QA: 

+ [ ] Code review
+ [ ] Test service upload UI on Internet Explorer
+ [ ] Test record/discontinue upload UI on Internet Explorer
+ [ ] Test Student Report PDF rendering on Internet Explorer

Jngai commented:
@alexsoble Thoughtbot is good stuff. Plus the deploy script tells me that the prod is on heroku. Honestly is just a `heroku run rake services:move_discontinued_at_to_services ` instead of a `heroku run rake db:migrate` if you have prod access and do it manually.

Jngai commented:
Also @alexsoble check if all info `discontinued_at` and `discontinued_by_educator_id` are collected correctly. I checked at the service creation process though.

alexsoble commented:
@Jngai In terms of the merge/deploy plan, this is a big change to the database. To deploy this, I would first make a manual backup of the database in case anything goes wrong. Next I would deploy to Heroku, run the migration, and watch how the migration runs. Finally, I would QA in production to make sure it ran correctly. The best time to do all of this would be in the evening after the school day is over to make sure no users are on the site while this process is happening. 

Since we've already planned another big data migration to run tonight (#1178 / https://github.com/studentinsights/studentinsights/pull/1178#issuecomment-339080724), let's aim for me to deploy this tomorrow evening around 7PM EST assuming all the reviews and manual testing run smoothly. 

Thanks for your patience on this and for dealing with the occasional merge conflicts that have cropped up!

alexsoble commented on app/controllers/service_uploads_controller.rb:
```diff
@@ -45,15 +45,9 @@ def create
-        if discontinued_service.valid?
-          discontinued_service.save
-        else
+        unless service.update_attributes(:discontinued_at => date_ended, :discontinued_by_educator_id => current_educator.id)
```
Why does this repeat the code on line 48? We should only need to `update_attributes` once.

alexsoble replied:
So could we safely remove line 48? :D

alexsoble replied:
Happy to do some pairing on this together @Jngai!

alexsoble commented on app/controllers/services_controller.rb:
```diff
@@ -10,18 +10,15 @@ def authorize!
+    serializer = ServiceSerializer.new(service)
 
-    if discontinued_service.save
+    if service.update_attributes(:discontinued_at => Time.now, :discontinued_by_educator_id => current_educator.id)
```
Looks like we're updating attributes twice here too.

alexsoble commented on app/serializers/service_serializer.rb:
```diff
@@ -11,13 +11,13 @@ def serialize_service
     })
   end
 
   def discontinued
```
Looks like we don't need this method anymore.

alexsoble commented on spec/controllers/students_controller_spec.rb:
```diff
@@ -450,7 +450,7 @@ def make_request
       expect(feed.keys).to eq([:event_notes, :services, :deprecated])
       expect(feed[:services].keys).to eq [:active, :discontinued]
-      expect(feed[:services][:active].first[:id]).to eq service.id
+      expect(feed[:services][:discontinued].first[:id]).to eq service.id
```
What does this change mean / refer to?

alexsoble commented:
@Jngai I ran a code review and left comments / change requests in a few places.

Jngai commented:
@alexsoble I am completely out of hours so this is as far as I can get. That serializer stuff if I removed `merge` then it breaks and there is actually no way to make two factories like earlier what we did. One factory maps to one model only. This slightly change for #1200. For #1200 please only merge in `86840ec` thanks.

alexsoble commented:
OK @Jngai thank you for your work on this! I'll keep going with the QA and if all's clear will aim to merge/deploy tomorrow evening. On the hours, please email @snoopyuri and check on that status of your last invoice. Thanks again @Jngai!

Jngai commented:
@alexsoble This is a very big complicated pr. It should have gotten a couple more hours however I still think we are at a good state. I will email @snoopyuri tonight thank you.

Jngai commented:
@alexsoble I know you are merging this tonight. I will be around if you need help. I will be around if there are questions.

alexsoble commented:
Apologies @Jngai. I started feeling sick tonight so I'm not able to merge/deploy now.

Jngai commented:
@alexsoble no worries. Please merge my work when you can. I don't want to leave it around and let it be stale.

alexsoble commented:
@Jngai Sure thing. I wouldn't worry to much about it getting stale though, if other branches get merged in that include conflicts we can just run 

```
git rebase master
``` 

and that will pull in the latest changes without too much hassle. 

I'll be happy to do the rebasing myself if this PR falls behind master, or if you'd rather do it I'm happy to do another pairing session with you on rebasing!

Jngai commented:
@alexsoble Its not about the pr getting stale. Its about my work done, useful and not used. I did get a renewal. I will be happy to finish up that final test, model refactoring, and stand by when merged. You guys are paying for this. My checks are just coming in late but I am still getting it. Lets talk next week.

alexsoble commented:
Hi @Jngai! This PR is useful and will make the internal API for services simpler. We also have to look at what else is going on in the project. 

Last week we had two time-sensitive events to handle: a new format of MCAS test data flowed into the project (which needed new code to handle the new format), and we had a short window to import a new batch of IEP data to backfill a gap. 

Because this product is serving real educators, it's important for us to focus on merging in PRs that fix data issues or add missing data first. PRs that refactor data structures will be merged in later. It's extra important that we take our time to review and carefully QA PRs that make potentially changes to data. 

Hope this explanation about when and why we merge PRs into the project is helpful! Feel free to add more questions. This is great work, but we can't rush or hurry this PR because it's a change that is developer-facing only and includes a highly sensitive migration.

alexsoble commented:
@Jngai I added a checklist with "steps-toward-merge" at the top level of this PR. 

I also added a commit which ships the simplification in `service.rb` that we talked about in our pairing session last week. 

I think this is important to include in this PR because without it this PR would leave `service.rb` with some confusing and "stubby" methods that create indirection: for example, methods like `def end_date; discontinued_at end`.

alexsoble commented on scripts/deploy/deploy.sh:
```diff
@@ -26,3 +26,11 @@ echo;
 heroku run rake db:migrate --app "$DEMO_HEROKU_APP_NAME"
 echo;
+
+echo "⚙  ⚙  ⚙  Running Service Rake Task for Demo App.";
```
@Jngai Just noticed this change. We don't want to add these lines to `deploy.sh` because  `deploy.sh` is meant to be used for every single deploy. We don't want to run the service migration on every single deploy, just once.

alexsoble replied:
@Jngai Sorry for not noticing this issue on my first review!

Jngai replied:
@alexsoble I thought I make it easier for you. I removed them so you have to run the commands manually.

Jngai commented:
@alexsoble I checked your last two commits into `service` model and I think they are good. Unless you need me I think the checklist is on your court. Please ping me if you need me.

alexsoble commented:
@Jngai While running through the QA for this PR I found that this breaks the Student Report feature by making it look as if discontinued services have never been discontinued. 

Take a look at line 93 of `app/views/students/student_report.pdf.erb` to see the issue here @Jngai. 

In addition to updating `student_report.pdf.erb`, could you write a test case for this to make sure this area of the student report has test coverage? In addition, could you do a grep or Sublime search for `discontinued_services` to make sure we're not overlooking any other areas that are in need of updating?

Jngai commented:
@alexsoble I added the bug fix but its not printing for me yet. Let me take a look.

alexsoble commented:
@Jngai OK. How do you feel about writing the associated spec as well? Does that seem doable or potentially challenging?

Jngai commented:
@alexsoble sublime search for `discontinued_service` looks good.

Jngai commented:
@alexsoble I tried removing `<% if @sections.include?("services") %>` in `app/views/students/student_report.pdf.erb` and the pdf started to print service stuff. 

<img width="1062" alt="screen shot 2017-11-02 at 2 20 00 pm" src="https://user-images.githubusercontent.com/1547934/32343178-0e80bcd2-bfd9-11e7-8a7d-738481ba944f.png">

Should pdf be printing services if services exist? The emptiness of `@sections` is excluding the service printing.

alexsoble commented:
@Jngai Sections are set by params:

```
# students_controller.rb

197   @sections = (params[:sections] || "").split(",")
```

The params are generated by the `onClickGenerateStudentReport` function in `profile_details.jsx`.

Does that help answer your question about `@sections` @Jngai?

Jngai commented:
@alexsoble That answers the question. There is a service test in students controller for student reports but I think you are looking for a view test that test the rendering of the pdf. I couldn't find any view tests. I also checked `profile_details_spec.jsx.`

alexsoble commented:
Hey there @Jngai! You are totally right, there aren't any view tests or integration tests for the Student Report. That was surprising to me, I assumed there would be. We shouldn't make that a blocker for shipping this PR. 

My plan is to re-run the QA checklist, merge, and ship this PR after school hours (3pm CT / 4pm ET) on Wednesday 11/8. I'm excited to get this merged! 👍

Jngai commented:
Great @alexsoble I was checking to see if I could add a test but that would be adding a brand new javascript test.

Jngai commented:
@alexsoble I will be around today if you need my help. I will be online.

alexsoble commented:
@Jngai 👍 , re-QA'ing now!

Jngai commented:
Great @alexsoble Thank you. Also I put #1180 on hold because this wasn't shipped yet.

alexsoble commented:
@Jngai In my review I found that the end dates of discontinued services are still not being printed out in the Student Report. Here's a commit that fixes this issue: cccc991b4d5ff0c7501455ae33b3607b244a1d64.

Jngai commented:
@alexsoble I notice you changed `recorded_at` to `discontinued_at`. The services section printed for me but I am not very familiar with this feature. If this fixed the issue then great.

Jngai commented:
@alexsoble Actually it did print for me, please see last pic I attached above. However if you want `discontinued_at` now then go for it.

alexsoble commented:
@Jngai I see why this is confusing — it looks like there is an active bug in the code. The goal of this code as I understand it is to show "{start date} - {end date}" for a service. But the code doesn't do that right now. I'm going to make a PR to fix this bug and get a second pair of eyes on it from @jhilde.

alexsoble commented:
@Jngai —  The goal of this code is to show "{start date} - {end date}" for a service.

The existing (non-flattened) code looks like this: 

```
<% if service.discontinued_services.count > 0%>
  <%= "#{format_date(service.discontinued_services.last.try(:recorded_at))}" %>
<% else %>
...
```

The code you wrote changed this section to:

```
<% if service.discontinued_at.present?%>
  <%= "#{format_date(service.try(:recorded_at))}" %>
<% else %>
...
```

What changes here is that instead of showing when the service discontinued, now the PR shows when the service was recorded. Does that make sense?

alexsoble commented:
@Jngai I'm going to turn to this again at 3PM CT/4PM ET today, if you're able to fix up the student report on this branch before then that will be great, otherwise I'll add the commit I made yesterday and use git cherry-pick to QA/merge with that.

Jngai commented:
@alexsoble I was in master and for the non flattened code `<%= "#{format_date(service.discontinued_services.last.try(:recorded_at))}" %>` there is no `recorded_at` for `discontinued_services`. This has always been an active bug in the code. I think your commit does look good and I thought you added to my branch last night. I will add the commit but most likely with your solution.

Jngai commented:
@alexsoble I pretty much added your solution. So "{start date} - {end date}" will be `service.date_started` and `service.try(:discontinued_at)`. Feel free to turn back again today.

alexsoble commented:
Thanks @Jngai!

alexsoble commented:
Jumped on a call with @Jngai — we confirmed there is a live bug in production here. Escalating this now in an email. After I send the email I will run the QA one more time and :crossed_fingers: merge this PR (which will fix the bug), deploy, run the associated migrations, report on the outcome.

alexsoble commented:
QA passed. Capturing a manual database backup now.

alexsoble commented:
Backup captured! Let's deploy this thing! 👍  @Jngai

alexsoble commented:
The migration and the data migration both ran without errors, so that's 👍 .

At first I was concerned about a data error after checking the state of the database afterward because some of the numbers didn't match up well. Then I discovered this was because a large number of DiscontinuedService records didn't have a `service_id`. After some productive digging through the data with @kevinrobinson, we determined the mostly likely reason for these records was Service Uploads that had been deleted, deleting their associated services through a `dependent: destroy`. We identified several improvements we can make here which I'll record as separate issues.

Overall, 👍 🌈  work @Jngai, it feels good that we've got this merged and deployed!
